rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,94 @@
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+Automatic updates for client tools can be configured through tsh configuration. If `DISABLE_AUTO_UPDATE=true`, then auto updates for client tools will be disabled. If `DISABLE_UPDATE_PROMPT=true`, then the user will not be prompted to confirm the update. Both values will default to false.
+
+## UX
+
+Ideally, we'd like this feature to be integrated seamlessly. Users of the Teleport client tools should not need to search for documentation and spend time figuring out how to enable auto updates for their cleint tools.
+
+After client tools auto updates is deployed and the first time the client tools are executed, the user will be asked two questions: 1) if they would like to update to the latest version (default yes) and 2) if they would like to enroll in auto updates for client tools (default yes).
+
+Users that would like to opt-in or opt-out of auto updates at a later time can edit their tsh configuration file and edit the `DISABLE_AUTO_UPDATE` value.
+
+If a user opts out of auto updates, the user will still be prompted with the first question every time an update is available, but not the second question.
+
+We should provide some observability into the download status. Whenever a new cloud-stable version is available for download, the progress of the update will be output to stdout. e.g. `New cloud-stable version of Teleport detected`, `Downloading latest version of tsh/tctl... `, `Updated tsh/tctl to version v13.2.3!`.
+
+To avoid breaking existing scripts, a `--skip-auto-update` flag will be included. If the flag is enabled, auto updates will be skipped and the prompts will also be skipped.
+
+## Documentation
+
+There should be minimal documentation changes required for this feature. Users will not need to opt-in or take any actions to take advantage of this feature.
+
+The documentation should include a basic overview of how auto updates works for client tools. Documention should also be available describing how to use the configuration values `DISABLE_AUTO_UPDATE`, `DISABLE_UPDATE_PROMPT`, and the `--skip-auto-update` flag.
+
+## Security Considerations
+
+All downloads will be verified using the provided sha256 checksums. As long as our CDN is secure, the update process should be secure.
```

Comment:
How do we provide the shasums? through the same way of distributing content? If a service account with CDN acess gets comproised, an attacker can send an arbitrary binary to all customers. This also makes the update process vulnerable to MITM attacks.

To mitigate this I think we should sign the binaries and have tctl/tsh validate the signature before finishing the update.

Another approach that can complement signature would be to send shasums in the Teleport ping response.

Reply:
Thank you @hugoShaka, I agree.  It was not mentioned above when discussing the use of package managers to distribute updates, but that is a primary benefit we get from that delivery method.  Package managers help solve the supply chain security.  With this solution we now must fully take that burden on.

Hashes can be used to verify the binary is as expected and trusted, but we must consider how those hashes are communicated. It sounds like the current thought is to use TLS Certificate verification to make sure we are talking to our CDN for those hashes.  Is that correct?  Are those hashes then just retained in memory?  Some additional details around this process would be great.

Reply:
Hey, I'm not as familiar with security in software distribution so all suggestions are welcome. I'll add these as options for securing the update.

> How do we provide the shasums?

We provide checksums at `https://get.gravitational.com/teleport-ent-<version>-<os>-<arch>-bin.tar.gz.sha256`. I was thinking we can just pull the sha256 from here. 

> To mitigate this I think we should sign the binaries and have tctl/tsh validate the signature before finishing the update.

Would this be fairly straight forward to implement? Also would this allow the cached version of `tsh` to support TouchID on macOS?

> Are those hashes then just retained in memory?

Do we need to retain the hashes? I was thinking tsh/tctl would just pull the hash whenever they update.

Reply:
hashes are here to ensure integrity, signature ensures authenticity. If we ensure hash provenance with a signature, we can serve binaries, hashes, and signatures from the same CDN. Even if someone compromises the CDN, they won't be able to craft valid signatures and existing clients will refuse to update.

> Would this be fairly straight forward to implement? 

Implementing signature validation should be straightforward in an OS-independent way. (You hash the binary, then sign the hash. The downloader will download the hash, check its signature against a static keychain, then download the binary and check it matches the hash. This is what happens for docker images under the hood, you end up with both integrity and authenticity).

> Also would this allow the cached version of tsh to support TouchID on macOS?

The OS-agnostic way of singing is different from OS-specific notarization. I don't know much about MacOS TouchID requirements, but this will likely involve Mac-specific binary notarization, which differs from what I described above.

> Do we need to retain the hashes?

It depends what your threat model is. If you want to protect against a malicious CDN, a MiTM attack, or network glitches, then you only need to keep the hash to validate the binary during the update.

If you want to protect against bit rot or an attacker tampering with the already-downloaded binary, keep the hash to check for every invocation. However, as teleport is not a long-running process, we don't have a way other than to write it to the disk. Someone with enough rights to tamper the tsh binary will also have rights to tamper the hash, so we'd need to keep the signature as well. This kind of risk is usually mitigated by OS-notarization (windows or MacOS will do the check for you).

We can either store the hash and the signature on disk and validate both on each startup (I am afraid about the performance cost for each invocation) or consider this out of scope. We are already vulnerable to local tampering, the main thing that changes with the new auto-update mechanism is that you don't have to be root to tamper tsh, you can be the user.

IMO someone with enough rights to write will likely have enough rights to read and would be able to exfiltrate certs anyway. @jentfoo WDYT?

Reply:
I agree with everything you outlined above.  If we want to protect against tampering we have to move the source for validation off disk, and that is challenging with the short process lifecycle.

If we wanted to protect against tampering we would need to have a long running process which can securely load the hash list into memory on startup.  Then the cli would either communicate with it or be executed through this service to ensure validity.

I agree this does not have to be a requirement of this RFD, I don't believe moving the executable under the users control significantly increases any local attacks.  But we should still make sure the supply chain is as secure as possible.  I think @hugoShaka's suggestions above around signing the hashes and binary are excellent improvements.

Reply:
Okay, I've added signing/verifying the binaries as a security requirement. Protection against tampering is mentioned, but it will be left out of scope for now.